crates/router/src/utils/user.rs

```
@@ -339,3 +342,48 @@ pub async fn validate_email_domain_auth_type_using_db(
     .then_some(())
     .ok_or(UserErrors::InvalidUserAuthMethodOperation.into())
 }
+
+pub async fn set_lineage_context_in_cache(
+    state: &SessionState,
+    lineage_context: user_api::LineageContext,
+    user_id: String,
+) -> UserResult<()> {
+    let connection = get_redis_connection(state)?;
+    let key = format!("{}{}", LINEAGE_CONTEXT_PREFIX, user_id);
+    connection
+        .set_key_with_expiry(
+            &key.into(),
+            serde_json::to_string(&lineage_context)
+                .change_context(UserErrors::InternalServerError)
+                .attach_printable("Failed to serialize LineageContext")?,
+            LINEAGE_CONTEXT_TIME_EXPIRY_IN_SECS,
+        )
+        .await
+        .change_context(UserErrors::InternalServerError)
+        .attach_printable("Failed to set lineage context in redis")?;
+    Ok(())
+}
+
+pub async fn get_lineage_context_from_cache(
+    state: &SessionState,
+    user_id: String,
+) -> UserResult<Option<user_api::LineageContext>> {
+    let connection = get_redis_connection(state)?;
+    let key = format!("{}{}", LINEAGE_CONTEXT_PREFIX, user_id);
+    let lineage_context = connection
+        .get_key::<Option<String>>(&key.into())
+        .await
+        .change_context(UserErrors::InternalServerError)
+        .attach_printable("Failed to get lineage context from redis")?;
+
+    match lineage_context {
+        Some(json_str) => match serde_json::from_str::<user_api::LineageContext>(&json_str) {
+            Ok(ctx) => Ok(Some(ctx)),
+            Err(err) => {
+                logger::error!("Failed to deserialize LineageContext: {:?}", err);
+                Ok(None)
+            }
```

If we really don't want this to mess up our flows, we should throw all the errors here like we did for connection failure above and the callers should log and ignore the errors, which we are already doing. So, we can throw the actual error here and there's no need to fallback.

---

Though the helper functions are throwing errors, the error gets propagated and eventually logged, whenever the set_lineage_context and get_lineage_context are called.

We do not mean to disrupt the login and switch functionalities because of this feature